File: packages/kotlin/src/main/kotlin/me/really/crypto/proto/ProtoAdapterSupport.kt
// SPDX-FileCopyrightText: 2026 ReallyMe LLC
//
// SPDX-License-Identifier: MIT OR Apache-2.0

package me.really.crypto.proto

import com.google.protobuf.ByteString
import com.google.protobuf.InvalidProtocolBufferException
import me.really.crypto.ReallyMeAeadAlgorithm
import me.really.crypto.ReallyMeCryptoException
import me.really.crypto.ReallyMeHashAlgorithm
import me.really.crypto.ReallyMeHpkeSuite
import me.really.crypto.ReallyMeHpkeSealedMessage
import me.really.crypto.ReallyMeJwk
import me.really.crypto.ReallyMeJwkAlgorithm
import me.really.crypto.ReallyMeJwkKey
import me.really.crypto.ReallyMeKdfAlgorithm
import me.really.crypto.ReallyMeKemAlgorithm
import me.really.crypto.ReallyMeKemEncapsulation
import me.really.crypto.ReallyMeKemKeyPair
import me.really.crypto.ReallyMeKeyAgreementKeyPair
import me.really.crypto.ReallyMeKeyAgreementAlgorithm
import me.really.crypto.ReallyMeKeyWrapAlgorithm
import me.really.crypto.ReallyMeMacAlgorithm
import me.really.crypto.ReallyMeNativeStatus
import me.really.crypto.ReallyMeMulticodecKeyAlgorithm
import me.really.crypto.ReallyMeSignatureKeyPair
import me.really.crypto.ReallyMeSignatureAlgorithm
import me.really.crypto.v1.AeadAlgorithm
import me.really.crypto.v1.CryptoAlgorithmFamily
import me.really.crypto.v1.CryptoAlgorithmIdentifier
import me.really.crypto.v1.CryptoBackendError
import me.really.crypto.v1.CryptoError
import me.really.crypto.v1.CryptoErrorReason
import me.really.crypto.v1.CryptoHpkeSealedMessage
import me.really.crypto.v1.CryptoKemEncapsulation
import me.really.crypto.v1.CryptoKeyPair
import me.really.crypto.v1.CryptoProviderCapability
import me.really.crypto.v1.CryptoProviderCapabilitySet
import me.really.crypto.v1.CryptoPrimitiveError
import me.really.crypto.v1.CryptoProviderError
import me.really.crypto.v1.CryptoProviderSupportStatus
import me.really.crypto.v1.CryptoVerificationResult
import me.really.crypto.v1.CryptoVerificationStatus
import me.really.crypto.v1.HashAlgorithm
import me.really.crypto.v1.HpkeAeadId
import me.really.crypto.v1.HpkeKdfId
import me.really.crypto.v1.HpkeKemId
import me.really.crypto.v1.HpkeSuiteIdentifier
import me.really.crypto.v1.JsonWebKey
import me.really.crypto.v1.JsonWebKeySet
import me.really.crypto.v1.KdfAlgorithm
import me.really.crypto.v1.KemAlgorithm
import me.really.crypto.v1.KeyAgreementAlgorithm
import me.really.crypto.v1.KeyWrapAlgorithm
import me.really.crypto.v1.MacAlgorithm
import me.really.crypto.v1.MulticodecKeyAlgorithm
import me.really.crypto.v1.SignatureAlgorithm

internal fun jwkAlgorithmToProto(
    value: ReallyMeJwkAlgorithm,
): CryptoAlgorithmIdentifier =
    when (value) {
        ReallyMeJwkAlgorithm.ED25519 -> CryptoAlgorithmIdentifier.newBuilder()
            .setSignature(SignatureAlgorithm.SIGNATURE_ALGORITHM_ED25519)
            .build()
        ReallyMeJwkAlgorithm.X25519 -> CryptoAlgorithmIdentifier.newBuilder()
            .setKeyAgreement(KeyAgreementAlgorithm.KEY_AGREEMENT_ALGORITHM_X25519)
            .build()
        ReallyMeJwkAlgorithm.P256 -> CryptoAlgorithmIdentifier.newBuilder()
            .setSignature(SignatureAlgorithm.SIGNATURE_ALGORITHM_ECDSA_P256_SHA256)
            .build()
        ReallyMeJwkAlgorithm.SECP256K1 -> CryptoAlgorithmIdentifier.newBuilder()
            .setSignature(SignatureAlgorithm.SIGNATURE_ALGORITHM_ECDSA_SECP256K1_SHA256)
            .build()
        ReallyMeJwkAlgorithm.ML_DSA_44 -> CryptoAlgorithmIdentifier.newBuilder()
            .setSignature(SignatureAlgorithm.SIGNATURE_ALGORITHM_ML_DSA_44)
            .build()
        ReallyMeJwkAlgorithm.ML_DSA_65 -> CryptoAlgorithmIdentifier.newBuilder()
            .setSignature(SignatureAlgorithm.SIGNATURE_ALGORITHM_ML_DSA_65)
            .build()
        ReallyMeJwkAlgorithm.ML_DSA_87 -> CryptoAlgorithmIdentifier.newBuilder()
            .setSignature(SignatureAlgorithm.SIGNATURE_ALGORITHM_ML_DSA_87)
            .build()
        ReallyMeJwkAlgorithm.ML_KEM_512 -> CryptoAlgorithmIdentifier.newBuilder()
            .setKem(KemAlgorithm.KEM_ALGORITHM_ML_KEM_512)
            .build()
        ReallyMeJwkAlgorithm.ML_KEM_768 -> CryptoAlgorithmIdentifier.newBuilder()
            .setKem(KemAlgorithm.KEM_ALGORITHM_ML_KEM_768)
            .build()
        ReallyMeJwkAlgorithm.ML_KEM_1024 -> CryptoAlgorithmIdentifier.newBuilder()
            .setKem(KemAlgorithm.KEM_ALGORITHM_ML_KEM_1024)
            .build()
        ReallyMeJwkAlgorithm.SLH_DSA_SHA2_128S -> CryptoAlgorithmIdentifier.newBuilder()
            .setSignature(SignatureAlgorithm.SIGNATURE_ALGORITHM_SLH_DSA_SHA2_128S)
            .build()
        ReallyMeJwkAlgorithm.X_WING_768 -> CryptoAlgorithmIdentifier.newBuilder()
            .setKem(KemAlgorithm.KEM_ALGORITHM_X_WING_768)
            .build()
    }

internal fun jwkAlgorithmFromProto(
    value: CryptoAlgorithmIdentifier,
): ReallyMeJwkAlgorithm =
    when (value.algorithmCase) {
        CryptoAlgorithmIdentifier.AlgorithmCase.SIGNATURE -> when (value.signature) {
            SignatureAlgorithm.SIGNATURE_ALGORITHM_ED25519 ->
                ReallyMeJwkAlgorithm.ED25519
            SignatureAlgorithm.SIGNATURE_ALGORITHM_ECDSA_P256_SHA256 ->
                ReallyMeJwkAlgorithm.P256
            SignatureAlgorithm.SIGNATURE_ALGORITHM_ECDSA_SECP256K1_SHA256 ->
                ReallyMeJwkAlgorithm.SECP256K1
            SignatureAlgorithm.SIGNATURE_ALGORITHM_ML_DSA_44 ->
                ReallyMeJwkAlgorithm.ML_DSA_44
            SignatureAlgorithm.SIGNATURE_ALGORITHM_ML_DSA_65 ->
                ReallyMeJwkAlgorithm.ML_DSA_65
            SignatureAlgorithm.SIGNATURE_ALGORITHM_ML_DSA_87 ->
                ReallyMeJwkAlgorithm.ML_DSA_87
            SignatureAlgorithm.SIGNATURE_ALGORITHM_SLH_DSA_SHA2_128S ->
                ReallyMeJwkAlgorithm.SLH_DSA_SHA2_128S
            else -> throw ReallyMeCryptoException.UnsupportedAlgorithm()
        }
        CryptoAlgorithmIdentifier.AlgorithmCase.KEY_AGREEMENT -> {
            if (value.keyAgreement == KeyAgreementAlgorithm.KEY_AGREEMENT_ALGORITHM_X25519) {
                ReallyMeJwkAlgorithm.X25519
            } else {
                throw ReallyMeCryptoException.UnsupportedAlgorithm()
            }
        }
        CryptoAlgorithmIdentifier.AlgorithmCase.KEM -> when (value.kem) {
            KemAlgorithm.KEM_ALGORITHM_ML_KEM_512 -> ReallyMeJwkAlgorithm.ML_KEM_512
            KemAlgorithm.KEM_ALGORITHM_ML_KEM_768 -> ReallyMeJwkAlgorithm.ML_KEM_768
            KemAlgorithm.KEM_ALGORITHM_ML_KEM_1024 -> ReallyMeJwkAlgorithm.ML_KEM_1024
            KemAlgorithm.KEM_ALGORITHM_X_WING_768 -> ReallyMeJwkAlgorithm.X_WING_768
            else -> throw ReallyMeCryptoException.UnsupportedAlgorithm()
        }
        CryptoAlgorithmIdentifier.AlgorithmCase.ALGORITHM_NOT_SET ->
            throw ReallyMeCryptoException.InvalidInput()
        else -> throw ReallyMeCryptoException.UnsupportedAlgorithm()
    }

internal fun keyPairToProto(
    algorithm: CryptoAlgorithmIdentifier,
    publicKey: ByteArray,
    secretKey: ByteArray,
): CryptoKeyPair =
    CryptoKeyPair.newBuilder()
        .setAlgorithm(algorithm)
        .setPublicKey(ByteString.copyFrom(publicKey))
        .setSecretKey(ByteString.copyFrom(secretKey))
        .build()

internal fun signatureAlgorithmIdentifierToProto(
    value: ReallyMeSignatureAlgorithm,
): CryptoAlgorithmIdentifier =
    CryptoAlgorithmIdentifier.newBuilder()
        .setSignature(ReallyMeCryptoProtoAdapters.toProto(value))
        .build()

internal fun keyAgreementAlgorithmIdentifierToProto(
    value: ReallyMeKeyAgreementAlgorithm,
): CryptoAlgorithmIdentifier =
    CryptoAlgorithmIdentifier.newBuilder()
        .setKeyAgreement(ReallyMeCryptoProtoAdapters.toProto(value))
        .build()

internal fun kemAlgorithmIdentifierToProto(
    value: ReallyMeKemAlgorithm,
): CryptoAlgorithmIdentifier =
    CryptoAlgorithmIdentifier.newBuilder()
        .setKem(ReallyMeCryptoProtoAdapters.toProto(value))
        .build()

internal fun hpkeSuiteIdentifierToProto(
    value: ReallyMeHpkeSuite,
): CryptoAlgorithmIdentifier =
    CryptoAlgorithmIdentifier.newBuilder()
        .setHpkeSuite(ReallyMeCryptoProtoAdapters.toProto(value))
        .build()

internal fun signatureAlgorithmFromIdentifier(
    value: CryptoAlgorithmIdentifier,
    isPresent: Boolean,
): ReallyMeSignatureAlgorithm {
    if (!isPresent || value.algorithmCase != CryptoAlgorithmIdentifier.AlgorithmCase.SIGNATURE) {
        throw ReallyMeCryptoException.InvalidInput()
    }
    return ReallyMeCryptoProtoAdapters.fromProto(value.signature)
}

internal fun keyAgreementAlgorithmFromIdentifier(
    value: CryptoAlgorithmIdentifier,
    isPresent: Boolean,
): ReallyMeKeyAgreementAlgorithm {
    if (!isPresent || value.algorithmCase != CryptoAlgorithmIdentifier.AlgorithmCase.KEY_AGREEMENT) {
        throw ReallyMeCryptoException.InvalidInput()
    }
    return ReallyMeCryptoProtoAdapters.fromProto(value.keyAgreement)
}

internal fun kemAlgorithmFromIdentifier(
    value: CryptoAlgorithmIdentifier,
    isPresent: Boolean,
): ReallyMeKemAlgorithm {
    if (!isPresent || value.algorithmCase != CryptoAlgorithmIdentifier.AlgorithmCase.KEM) {
        throw ReallyMeCryptoException.InvalidInput()
    }
    return ReallyMeCryptoProtoAdapters.fromProto(value.kem)
}

internal fun hpkeSuiteFromIdentifier(
    value: CryptoAlgorithmIdentifier,
    isPresent: Boolean,
): ReallyMeHpkeSuite {
    if (!isPresent || value.algorithmCase != CryptoAlgorithmIdentifier.AlgorithmCase.HPKE_SUITE) {
        throw ReallyMeCryptoException.InvalidInput()
    }
    return ReallyMeCryptoProtoAdapters.fromProto(value.hpkeSuite)
}

File: packages/swift/Sources/ReallyMeCryptoProtoAdapters/ProtoAdapterSupport.swift
// SPDX-FileCopyrightText: 2026 ReallyMe LLC
//
// SPDX-License-Identifier: MIT OR Apache-2.0

import Foundation
import ReallyMeCrypto
import ReallyMeCryptoProto
import SwiftProtobuf

extension ReallyMeCryptoProtoAdapters {
  static func serialized<T: SwiftProtobuf.Message>(_ value: T) throws(ReallyMeCryptoError)
    -> [UInt8]
  {
    do {
      return try value.serializedBytes()
    } catch {
      throw ReallyMeCryptoError.providerFailure
    }
  }

  static func keyPairToProto(
    algorithm: ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier,
    publicKey: [UInt8],
    secretKey: [UInt8]
  ) -> ReallyMeCryptoProto.ReallyMeProtoCryptoKeyPair {
    var proto = ReallyMeCryptoProto.ReallyMeProtoCryptoKeyPair()
    proto.algorithm = algorithm
    proto.publicKey = Data(publicKey)
    proto.secretKey = Data(secretKey)
    return proto
  }

  static func signatureAlgorithmIdentifierToProto(
    _ value: ReallyMeSignatureAlgorithm
  ) -> ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier {
    var algorithm = ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier()
    algorithm.signature = toProto(value)
    return algorithm
  }

  static func keyAgreementAlgorithmIdentifierToProto(
    _ value: ReallyMeKeyAgreementAlgorithm
  ) -> ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier {
    var algorithm = ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier()
    algorithm.keyAgreement = toProto(value)
    return algorithm
  }

  static func kemAlgorithmIdentifierToProto(
    _ value: ReallyMeKemAlgorithm
  ) -> ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier {
    var algorithm = ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier()
    algorithm.kem = toProto(value)
    return algorithm
  }

  static func hpkeSuiteIdentifierToProto(
    _ value: ReallyMeHpkeSuite
  ) -> ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier {
    var algorithm = ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier()
    algorithm.hpkeSuite = toProto(value)
    return algorithm
  }

  static func signatureAlgorithm(
    fromIdentifier value: ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier,
    isPresent: Bool
  ) throws(ReallyMeCryptoError) -> ReallyMeSignatureAlgorithm {
    guard isPresent else {
      throw ReallyMeCryptoError.invalidInput
    }
    guard case .signature(let signature)? = value.algorithm else {
      throw ReallyMeCryptoError.invalidInput
    }
    return try fromProto(signature)
  }

  static func keyAgreementAlgorithm(
    fromIdentifier value: ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier,
    isPresent: Bool
  ) throws(ReallyMeCryptoError) -> ReallyMeKeyAgreementAlgorithm {
    guard isPresent else {
      throw ReallyMeCryptoError.invalidInput
    }
    guard case .keyAgreement(let keyAgreement)? = value.algorithm else {
      throw ReallyMeCryptoError.invalidInput
    }
    return try fromProto(keyAgreement)
  }

  static func kemAlgorithm(
    fromIdentifier value: ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier,
    isPresent: Bool
  ) throws(ReallyMeCryptoError) -> ReallyMeKemAlgorithm {
    guard isPresent else {
      throw ReallyMeCryptoError.invalidInput
    }
    guard case .kem(let kem)? = value.algorithm else {
      throw ReallyMeCryptoError.invalidInput
    }
    return try fromProto(kem)
  }

  static func hpkeSuite(
    fromIdentifier value: ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier,
    isPresent: Bool
  ) throws(ReallyMeCryptoError) -> ReallyMeHpkeSuite {
    guard isPresent else {
      throw ReallyMeCryptoError.invalidInput
    }
    guard case .hpkeSuite(let hpke)? = value.algorithm else {
      throw ReallyMeCryptoError.invalidInput
    }
    return try fromProto(hpke)
  }

  static func jwkAlgorithmToProto(
    _ value: ReallyMeJwkAlgorithm
  ) throws(ReallyMeCryptoError) -> ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier {
    var algorithm = ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier()
    switch value {
    case .ed25519:
      algorithm.signature = .ed25519
    case .x25519:
      algorithm.keyAgreement = .x25519
    case .p256:
      algorithm.signature = .ecdsaP256Sha256
    case .secp256k1:
      algorithm.signature = .ecdsaSecp256K1Sha256
    case .mlDsa44:
      algorithm.signature = .mlDsa44
    case .mlDsa65:
      algorithm.signature = .mlDsa65
    case .mlDsa87:
      algorithm.signature = .mlDsa87
    case .mlKem512:
      algorithm.kem = .mlKem512
    case .mlKem768:
      algorithm.kem = .mlKem768
    case .mlKem1024:
      algorithm.kem = .mlKem1024
    case .slhDsaSha2_128s:
      algorithm.signature = .slhDsaSha2128S
    case .xWing768:
      algorithm.kem = .xWing768
    }
    return algorithm
  }

  static func jwkAlgorithm(
    fromProto value: ReallyMeCryptoProto.ReallyMeProtoCryptoAlgorithmIdentifier
  ) throws(ReallyMeCryptoError) -> ReallyMeJwkAlgorithm {
    switch value.algorithm {
    case .signature(let signature):
      switch signature {
      case .ed25519:
        return .ed25519
      case .ecdsaP256Sha256:
        return .p256
      case .ecdsaSecp256K1Sha256:
        return .secp256k1
      case .mlDsa44:
        return .mlDsa44
      case .mlDsa65:
        return .mlDsa65
      case .mlDsa87:
        return .mlDsa87
      case .slhDsaSha2128S:
        return .slhDsaSha2_128s
      default:
        throw ReallyMeCryptoError.unsupportedAlgorithm
      }
    case .keyAgreement(let keyAgreement):
      guard keyAgreement == .x25519 else {
        throw ReallyMeCryptoError.unsupportedAlgorithm
      }
      return .x25519
    case .kem(let kem):
      switch kem {
      case .mlKem512:
        return .mlKem512
      case .mlKem768:
        return .mlKem768
      case .mlKem1024:
        return .mlKem1024
      case .xWing768:
        return .xWing768
      default:
        throw ReallyMeCryptoError.unsupportedAlgorithm
      }
    case nil:
      throw ReallyMeCryptoError.invalidInput
    default:
      throw ReallyMeCryptoError.unsupportedAlgorithm
    }
  }
}
